numpy/lib/arraysetops.py

```diff
@@ -196,6 +196,12 @@ def unique(ar, return_index=False, return_inverse=False, return_counts=False):
         aux = ar
     flag = np.concatenate(([True], aux[1:] != aux[:-1]))
 
+    # gh-2111: NaNs never compare equal, so they need special handling,
+    # but they always sort to the end
+    if issubclass(ar.dtype.type, np.inexact) and np.isnan(aux[-1]):
+        nanidx = np.searchsorted(aux, np.nan)
```

Comment:
what if you replace this with `nanidx = np.nonzero(np.isnan(aux))[0]` ?

Reply:
Well, that creates an intermediate array of the same size as `aux`, so it's O(n) in both space and time. The solution above is O(log n) in time and O(1) in space, so it will almost certainly have better performance. It could be argued that your proposed code may be more understandable to a beginner, but I don't think `np.searchsorted` should be considered advanced or obscure, and should be perfectly fine to use, especially in numpy's source code.

So I guess I like my code better... ;-)

Reply:
@mattip's solution has the benefit of working for all types. So maybe we want both?